public/serviceWorker.js

```diff
@@ -0,0 +1,366 @@
+     * and the fetch request timeouting at 300 seconds.
+     * If this happens we would run into the Automatic reuse detection triggering in the backend.
+     */
+    const controller = new AbortController();
+
+    refreshingTokensLock.lock(async () => {
+        // Aborting the fetch request.
+        controller.abort()
+        await localforage.setItem(ACCESS_TOKEN, ACCESS_TOKEN_CONFIRMED_UNAUTHORIZED);
+    });
+
+    let accessToken;
+    let refreshResponse = await fetch('/api/auth/refresh', {
+        method: 'POST',
+        signal: controller.signal
+    });
+    event.waitUntil(refreshResponse);
+
+    /**
+     * Refresh token is valid, and we have received a new access token
+     */
+    if (refreshResponse && refreshResponse.ok) {
+        let jsonResponse = await refreshResponse.json();
+        accessToken = jsonResponse.access_token;
+        await localforage.setItem(ACCESS_TOKEN, jsonResponse.accessToken);
+    }
+
+    /**
+     * Refresh token is invalid, set to ACCESS_TOKEN_CONFIRMED_UNAUTHORIZED
+     */
+    if (refreshResponse.status === 422) {
+        accessToken = ACCESS_TOKEN_CONFIRMED_UNAUTHORIZED;
+        await localforage.setItem(ACCESS_TOKEN, accessToken);
+    }
+
+    refreshingTokensLock.unlock();
+    return accessToken
+}
+
+/**
+ * Handle Authorizes
+ *
+ * @param {Event} event
+ * @return {Response}
+ */
+async function handleAuthorizes(event) {
+    let response = await fetch(event.request);
+    event.waitUntil(response);
+
+    if (response && response.ok) {
+        let clone = response.clone();
+        let jsonResponse = await clone.json();
+        await localforage.setItem(ACCESS_TOKEN, jsonResponse.accessToken);
+    }
+
+    return response;
+}
+
+/**
+ * Handle UnAuthorizes
+ *
+ * @param {Event} event
+ * @param {Request} modifiedRequest
+ * @return {Response}
+ */
+async function HandleUnAuthorizes(event, modifiedRequest) {
+    let response = await fetch(modifiedRequest);
+    event.waitUntil(response);
+
+    if (response && (response.type === 'opaqueredirect') || response.ok) {
+        await localforage.setItem(ACCESS_TOKEN, ACCESS_TOKEN_CONFIRMED_UNAUTHORIZED);
+    }
+
+    return response;
+}
+
+/**
+ * Add Grammatip-Authorization header to a request
+ *
+ * @param {Request} request
+ * @param {String} accessToken
+ * @return {Request}
+ */
+function addAuthHeaderToRequest(request, accessToken) {
+    const modifiedHeaders = new Headers(request.headers);
+    modifiedHeaders.append('Grammatip-Authorization', (accessToken));
+
+    const modifiedRequestInit = { headers: modifiedHeaders, mode: 'same-origin' };
+    return new Request(request, modifiedRequestInit);
+}
+
+/**
+ * Authorize urls is urls where a user can get authorized.
+ * Urls where the response is an accessToken.
+ *
+ * @param {URL} url
+ * @return {boolean}
+ */
+function isAuthorizeUrls(url) {
+    const urlsExact = [
+        '/api/auth/login',
+        '/api/auth/register',
+        '/api/auth/refresh',
+    ];
```

Comment:
Use PUBLIC_API_URL from .env

Reply:
Decided it was not worth it right now to add url.